rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+# SlateDB Merge Operator
+
+Status: Draft
+
+Authors:
+
+* [David Moravek](https://github.com/dmvk)
+
+References:
+
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity
+* https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go
+
+## Motivation
+
+To utilize SlateDB as a fully-featured state backend for a stream processor, it must support three key types of state:
+
+1. **Value**: Latest value or aggregating state.
+2. **List / Bag / Buffering**: State that stores multiple items for later processing.
+3. **Map**: Key-value state for fine-grained data.
+
+State-of-the-art implementations built on RocksDB often leverage the [Merge Operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator) to efficiently manage types 1 and 2. A prime example is list state, where elements are appended continuously until triggered by an event like a window firing. While this method may not always be the absolute optimal solution, it serves as a solid foundation for future iterations and enhancements.
+
+## Goals
+The primary goal is to introduce a new user-facing record type (**Merge Operand**) and a **Merge Operator** interface to SlateDB, allowing applications to bypass the traditional read/modify/update cycle in performance-critical situations where computation can be expressed using an [associative](https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity) operator.
+
+**Supporting non-associative operators is not a goal of this RFC**. The assumption is that associative operators can cover the majority of use cases while keeping the user-facing API simple and significantly reducing the scope. [Badger](https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go) has adopted the same limitation.
+
+## Public API
+
+We will introduce two new methods for writing Merge Operands into the database. They follow the same structure as existing put methods.
+
+```rust
+impl Db {
+    
+    pub async fn merge(
```

Comment:
I think we'd also want to support including a merge in a write batch, right? It seems useful especially since write batches are a foundational piece for transactions and it would be good to be able to include a merge in a transaction.

Reply:
There should be only one final merge operation for every key, right? Rest can be merged together as part of committing the batch.

Reply:
Depends on how the TTL discussion turns out. It's also possible to have multiple merges in case MergeOperator returns an error.

Reply:
I think how it's implemented is a different matter - we could merge the entries together or just write them separately - I'm just saying that we should support the operation in a write batch.